repoman/git/depot_operations.py

```diff
@@ -58,24 +58,32 @@ def grab_changesets(self, path, url, changesets):
         try:
             if git_repo.is_bare:
                 subprocess.call('git fetch %s' % url, cwd=path, shell=True)
-                #this second fetch is needed to set HEAD
+                # this second fetch is needed to set HEAD
                 subprocess.call('git fetch', cwd=path, shell=True)
                 logger.debug('GIT Done grabbing changesets from github')
             else:
-                previous_branch = git_repo.head.shorthand
-                subprocess.call('git reset --hard', cwd=path, shell=True)
+                previous_branch = None
+                if not git_repo.head_is_detached and not git_repo.head_is_unborn:
+                    previous_branch = git_repo.head.shorthand
+
                 subprocess.call('git checkout --detach', cwd=path, shell=True)
                 subprocess.call('git fetch origin', cwd=path, shell=True)
-                subprocess.call('git checkout %s' % previous_branch, cwd=path,
+
+                if not previous_branch:
+                    previous_branch = git_repo.listall_branches()[0]
```

Comment:
First branch happens to be the master branch?

Reply:
Well, it doesn't really matter what branch is checked out in this `grab_changesets` method, right?

Reply:
In this function, it doesn't matter as you said. 
It should just retrieve new references from upstream (if it is the main_cache) or from main_cache (if it is a workspace).